### ISSUE TITLE
Docker container (running in Kubernetes) does not persist apt packages installed

### ISSUE THREAD

**Issue by nodesocket** [cdr/code-server#2924]
## OS/Web Information

- Web Browser: Chrome 89.0.4389.90
- Local OS: macOS Big Sur
- Remote OS: Kubernetes using official Helm chart (https://github.com/cdr/code-server/tree/main/ci/helm-chart)
- Remote Architecture: x64 Linux
- `code-server --version`:

```bash
kubectl exec -it code-server-5b7d67d849-wtwsw -- /bin/bash
    coder@code-server-5b7d67d849-wtwsw:~$ code-server --version
        3.9.1 e0203f2a36c9b7036fefa50eec6cf8fa36c5c015
```

When I use Terminal in code-server to install packages such as `zsh` with `sudo apt-get install zsh` these changes are lost when I redeploy the Helm chart update with a new version of code-server. This seems related to not mounting a volume to where global files `/usr/bin, /usr/local/bin` would be installed and changed. As a temporary solution, would it be possible to pre-install zsh into the Docker container? I bet a lot of people use `zsh` by default and would appreciate it.

**Comment by nodesocket** [comment]
Looks like we can just add `zsh` right after `dumb-init` here: https://github.com/cdr/code-server/blob/main/ci/release-image/Dockerfile#L6

**Comment by oxy** [comment]
Hi!

It's expected behavior for packages to be reset whenever code-server is upgraded, as every update pulls a fresh Docker image. As a solution, I'd probably recommend creating a personalize script or similar, and then re-running that script on every update.

In addition, mounting volumes on `/usr/bin` is a relatively unstable solution, as sometimes packages might install to `/opt`, or install library dependencies to `/usr/lib` or similar.

When it comes to including `zsh` in our default Dockerfile, I'd defer to one of my teammates.

@jsjoeio thoughts on shipping zsh by default?

**Comment by code-asher** [comment]
My general sense is that it would be better to build from code-server's Docker image and add the software you want on top, or the personalize script as you mentioned.

Wouldn't changing the shell to zsh from bash require some setup anyway? If there's already setup it feels like zsh could (should?) be installed during that same setup.

**Comment by nodesocket** [comment]
@code-asher thanks for the reply, appreciate it.

I am not suggesting we change the default shell, just that `zsh` is pre-installed for users that wish to use `zsh` and it being installed persists across Docker container upgrades. All the default shell configuration should be in the `$HOME` directory so that all persistists already as it's mounted as a volume.

See my PR https://github.com/cdr/code-server/blob/main/ci/release-image/Dockerfile#L6

**Comment by code-asher** [comment]
Ah yeah I didn't mean to imply that we would be changing the default shell. I meant that if a consumer of code-server wants to change their default shell my initial reaction is that they should also be responsible for installing said shell (whether that's zsh, fish, etc).

I think my concern is that we don't have a clear metric for determining what gets into the default image and what doesn't. For example, if someone asks for fish do we add it? Java? Etc.

What if we ran a script in `$HOME` if it exists on startup which you could use to install software or perform any other customization you might need?

Downside though is that it's not cached in the Docker layers. :thinking:

What about pushing a Docker image that inherits from code-server's and adds zsh?

All that said, I don't think our Docker users would mind having zsh added to the image even if they don't use it? So maybe we go ahead and do it.

**Comment by jsjoeio** [comment]
Closed via #2925 2925
